### DISCUSSION TITLE
compiling the front end and back end locally by following the docs

### DISCUSSION THREAD

**Discussion by ymolists** [zitadel/zitadel#2890] [Q&A]
Hello all

I want to first say thank you so much for doing this great amazing tool. I m trying to compile the code from master locally and i am having issues following the documentation. Here is the environment i have locally

```
$uname -a
Linux ymodesk 5.4.0-91-generic #102-Ubuntu SMP Fri Nov 5 16:31:28 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

$git sha 
f9bb5381f0d01127995758b2a513519af0607524

$docker version
Client:
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:50:40 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:55:09 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b638
 docker-init:
  Version:          0.19.0
```

```
$cd /src/zitadel/src/zitadel-master \
    && DOCKER_BUILDKIT=1 docker build --build-arg BUILDARCH=amd64 -f build/local/Dockerfile.keys --target copy-keys -o .keys . \
    && COMPOSE_DOCKER_CLI_BUILD=1 DOCKER_BUILDKIT=1 docker-compose -f ./build/local/docker-compose-local.yml --profile database --profile init-backend -p zitadel up


+ DOCKER_BUILDKIT=1
+ docker build --build-arg BUILDARCH=amd64 -f build/local/Dockerfile.keys --target copy-keys -o .keys .
[+] Building 0.1s (8/8) FINISHED                                                                             
 => [internal] load build definition from Dockerfile.keys                                               0.0s
 => => transferring dockerfile: 223B                                                                    0.0s
 => [internal] load .dockerignore                                                                       0.0s
 => => transferring context: 35B                                                                        0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                        0.0s
 => [gen-keys 1/3] FROM docker.io/library/alpine:latest                                                 0.0s
 => [internal] load build context                                                                       0.0s
 => => transferring context: 97B                                                                        0.0s
 => CACHED [gen-keys 2/3] COPY build/local/keys.sh keys.sh                                              0.0s
 => CACHED [gen-keys 3/3] RUN chmod +x /keys.sh                                                         0.0s
 => ERROR [copy-keys 1/1] COPY --from=gen-keys /.keys /.keys                                            0.0s
------
 > [copy-keys 1/1] COPY --from=gen-keys /.keys /.keys:
------
failed to compute cache key: "/.keys" not found: not found

```

notice how i had to add `--build-arg BUILDARCH=amd64` to the command in the docs. But even with that it still fails to compile. I cant even see where or how that .keys file is generated. up until thy point where it tries to call gen-keys i was able to follow the docs and each command completed successfully. I am wondering if i missed a step ?

Regards

**Comment by adlerhurst** [comment]
hi @ymolists 

thanks for your kind words.

I try to reproduce this behaviour and provide you feedback as soon as possible.

cheers

**Reply by adlerhurst** [reply]
@ymolists 
The problem is, that the environment variables are missing.

The following command generates the keys with the given env-file:
```bash
COMPOSE_DOCKER_CLI_BUILD=1 DOCKER_BUILDKIT=1 \
&& docker-compose -f ./build/local/docker-compose-local.yml up keys
```

**Reply by ymolists** [reply]
Oh great ! thank you so much for finding this !

Now, there is a maybe a bug in the [documentation here](https://github.com/caos/zitadel/blob/main/guides/development.md) then :-) I don't think that command is listed. Also that command should maybe come before the user is instructed to run the commands i ran.

**Reply by adlerhurst** [reply]
no problem :)

you are absolutely right i filed [an issue](https://github.com/caos/zitadel/issues/2891) to improve the local usage or at least update the guides. I will keep you posted

**Reply by ymolists** [reply]
finally got passed it !

`build/local/Dockerfile.keys` is supposed to have:
```
FROM alpine:latest AS gen-keys
COPY build/local/keys.sh keys.sh
RUN chmod +x /keys.sh
ENTRYPOINT [ "/keys.sh" ]

FROM scratch AS copy-keys
COPY /.keys /.keys
```

as opposed to what it is now 
```
FROM alpine:latest AS gen-keys
COPY build/local/keys.sh keys.sh
RUN chmod +x /keys.sh
ENTRYPOINT [ "/keys.sh" ]

FROM scratch AS copy-keys
COPY --from=gen-keys /.keys /.keys
```

i dont think gen-keys has that directory baked in it when it is building !

**Reply by adlerhurst** [reply]
hi @ymolists 

We just merged https://github.com/caos/zitadel/pull/2892 which should simplify docker compose.

Let me know if you have any feedback :)

Happy new year!

**Reply by ymolists** [reply]
Here is my feedback

* I am hoping that the docs will be one command that generates everything using one single docker compose file. I am sure this is possible but i am hoping that we can move in that direction. Somthing like `docker compose -f <fname> run setup` and boom it does all the necessary initialisation.
* Would it make sense to require [docker compose v2](https://github.com/docker/compose)  as opposed to docker-compose (the old way of doing things) which is supposed to be the defacto way of doing compose files moving forward. Its confusing but i think its worth considering
* Would it make sense to make sure the resulting files created by running docker compose file are always owned by the user and not root. Its super annoying if you need sudo in your dev environment when you want to remove those generated directories. I am on linux and all you need is pass user-id and group-id. I am not sure how that works in docker-machine (mac/windows) but worst case this could be a simple fix in linux.
*  I am thinking that i will create a PR and contribute those suggestions ( if they make sense) to simplify this whole process and hopefully get it merged !

**Reply by adlerhurst** [reply]
Thanks for your feedback. 

Have you checked out the quickstart.md? This is possibly what you meant with the onliner?

I absolutely agree that we should outline to use docker compose v2. 

The user owned files sound like a very cool addition. 

I would love to get a PR from you thank you very much 🙏

**Reply by ymolists** [reply]
@adlerhurst thank you for the suggestion ! I am back at this after the holidays .. finally :-)

I was able to run that successfuly but now i am trying to start the backend and front end locally. The good news is i think i am able to start the server via vscode ina debugger yaaay ! The bad news is i still cant get the console to connect to the back end successfully !

I have to say the `development.md` is really confusing :-( i find that it does not say what you have to do in short steps for one particular use case. It contains ifs in then documentation and that to me is very confusing. for example :

```
Used to set the client id of the console. This step is for local development. If you don't work with a local backend you have to set the client id manually.
```

The above is not clear to me when you are supposed to do that step. I think maybe a better approach would be to take two scenarios and document each one separately. The case where the user is using docker compose for both console/backend is already documented in the quick start. So maybe you can split `development.md` into two separate sections. The first steps could be detailing how to start the server locally. The other would be detailing how to start the console locally and connect to a running back end. The documentation can first start the server first then the console in that order to make it easier for a new comer to follow each step.

```
$COMPOSE_DOCKER_CLI_BUILD=1 DOCKER_BUILDKIT=1 \
&& docker compose -f ./build/local/docker-compose-local.yml --profile console-stub up --exit-code-from client-id
[+] Running 1/0
 ⠿ Container local-client-id-1  Created                                                                                                                                                  0.1s
Attaching to local-client-id-1
```

The above is what i get when i run the `Initialize the console` step but it never returns and gets stuck. Do i need the backend to be running before running that step ? what is required ? i did all the steps in the order they are documented but i think i am maybe missing something

Here is the sha i am using. 
```
$git sha 
e99b7f4972076d224368c973be64da125d4a1a77
```

**Comment by ymolists** [comment]
@adlerhurst I also tried the same thing with the latest released docker linux binaries and i am getting the same result

```
$docker version
Client:
 Version:           20.10.9
 API version:       1.41
 Go version:        go1.16.8
 Git commit:        c2ea9bc
 Built:             Mon Oct  4 16:03:22 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.9
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.8
  Git commit:       79ea9d3
  Built:            Mon Oct  4 16:07:30 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.4.11
  GitCommit:        5b46e404f6b9f661a205e28d59c982d3634148f8
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Reply by fforootd** [reply]
Interesting, what version of docker compose do you have? It works for sure with `2.2.1`.

Also does you docker config look more or less like this

```json
{
  "builder": {
    "gc": {
      "defaultKeepStorage": "20GB",
      "enabled": true
    }
  },
  "experimental": false,
  "features": {
    "buildkit": true
  }
}
```

**Reply by ymolists** [reply]
```
$docker compose version
Docker Compose version v2.2.2
```

Also i dont have a config file now under `$HOME/.docker/config.json` and i cant figure how to generate a default one using docker so far.
